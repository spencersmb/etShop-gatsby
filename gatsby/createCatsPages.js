const path = require("path")

module.exports = async ({ actions, graphql }) => {
  const { createPage } = actions
  const perPage = 10
  const allCats = {}
  const excludeItems = ["uncategorized"]
  let pageNumber = 1

  const fetchCatPages = async (variables) =>
    await graphql(GET_CATS, variables).then(async ({ data }) => {
      const {
        wpgraphql: {
          categories: {
            nodes
          }
        }
      } = data

      await Promise.all(nodes
        .filter(node => !excludeItems.includes(node.slug))
        .map((cat) => {
          const {
            supportQuestions: {
              pageInfo: { hasNextPage, endCursor, hasPreviousPage }
            }
          } = cat
          const questions = cat.supportQuestions.nodes
          const catPageTemplate = path.resolve(`./src/templates/supportCatListPage.tsx`)
          const catPagePath = `/support/category/${cat.slug}/page/${pageNumber}`
          // console.log("cat.supportQuestions.pageInfo", cat.supportQuestions.pageInfo)
          // console.log("slug", cat.slug)
          // console.log("ID", cat.id)
          // console.log("pageNumber", pageNumber)
          // console.log("hasNextPage", hasNextPage)
          // console.log("hasPrevPage", hasPreviousPage)

          // add item to be excluded if no next page
          if (!hasNextPage) {
            excludeItems.push(cat.slug)
          }
          const prevPages = allCats[cat.slug] && allCats[cat.slug].pages ? allCats[cat.slug].pages : {}
          // check if object exists first
          allCats[cat.slug] = {
            pages: {
              ...prevPages,
              [pageNumber]: {
                path: catPagePath,
                component: catPageTemplate,
                context: {
                  cat: cat.name,
                  catId: cat.id,
                  questions,
                  pageNumber,
                  hasNextPage,
                  hasPrevPage: hasPreviousPage
                }
              }
            }
          }

          if (hasNextPage) {
            pageNumber++
            return fetchCatPages({ first: perPage, after: endCursor })
          }
        }))
      return allCats
    })

  await fetchCatPages({ first: perPage, after: null }).then(catItems => {
    Object.keys(catItems).map((cat) => {
      const pages = Object.keys(catItems[cat].pages)
      pages.map(page => {
        console.log(`create Cat Page ${cat}`)
        console.log(`create Cat PageNumber ${catItems[cat].pages[page].context.pageNumber}`)
        createPage(catItems[cat].pages[page])
      })
    })
  })
}

const GET_CATS = `
  query GET_ALL_CATS($first:Int $after:String){
      wpgraphql{
        categories {
          nodes {
            id
            name
            slug
            supportQuestions(after: $after, first: $first) {
              pageInfo{
                  hasNextPage
                  hasPreviousPage
                  endCursor
              }
              nodes {
                  title
                  id
                  slug
                  excerpt
                  acfSupportQuestions{
                      popularity
                  }
              }
            }
          }
        }
      }
    }
  `
