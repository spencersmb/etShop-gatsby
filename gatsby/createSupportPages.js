const path = require(`path`)
module.exports = async ({ actions, graphql }) => {
  const supportQuery = `
  {
  wpgraphql {
    supportQuestions(where: {status: PUBLISH}, first: 100){
      nodes{
        date
        uri
        id
        slug
        title
        content
        featuredImage {
          altText
          caption
          sourceUrl
        }
      }
    }
  }
}
  `

  return new Promise((resolve, reject) => {
    graphql(supportQuery).then(results => {

      if (results.errors) {
        console.log(results.errors)
        reject(results.errors)
      }
      results.data.wpgraphql.supportQuestions.nodes.forEach(question => {
        actions.createPage({
          path: `/support/${question.slug}/`,
          component: path.resolve(`./src/templates/supportQuestion.tsx`),
          context: {
            // blogId: blog.id,
            ...question
          }
        })
        console.log(`Created page: /support/${question.slug}`)
      })
      resolve()
    })

  })
}
