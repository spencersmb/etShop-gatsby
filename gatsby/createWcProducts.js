const path = require("path")
const GET_PAGES = `
  query GET_WC_PAGES{
      allWcProduct{
        edges{
          node{
            slug
            name
          }
        }
      }
    }
  `
module.exports = async ({ actions, graphql }) => {
  const { createPage } = actions
  const allPages = []
  const fetchPages = async () =>
    await graphql(GET_PAGES).then(({ data }) => {
      const {
        allWcProduct: {
          edges
        }
      } = data
      edges.map(page => {
        allPages.push(page.node)
      })
      return allPages
    })

  await fetchPages().then(allPages => {
    allPages.map(page => {
      console.log(`create page: ${page.slug}`)
      createPage({
        path: `/products/${page.slug}`,
        component: path.resolve(`./src/templates/productPage.tsx`),
        context: {
          slug: page.slug
        }
      })
    })
  })

  // return new Promise((resolve, reject) => {
  //   graphql(pageQuery).then(results => {
  //
  //     if (results.errors) {
  //       console.log(results.errors)
  //       reject(results.errors)
  //     }
  //     results.data.allWcProduct.edges.forEach(({ node }) => {
  //       createPage({
  //         path: `/products/${node.slug}`,
  //         component: path.resolve(`./src/components/products/productDetailPage.tsx`),
  //         context: {
  //           slug: node.slug
  //         }
  //       })
  //       console.log(`Created page: /products/${node.slug}`)
  //     })
  //   })
  //   resolve()
  // })
}
