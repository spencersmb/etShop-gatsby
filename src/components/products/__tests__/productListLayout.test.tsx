import React from 'react'
import renderer from 'react-test-renderer'
import {
	render,
	cleanup
} from 'react-testing-library'
import 'jest-dom/extend-expect'
import ProductsListLayout from '../productsListLayout'
import { StaticQuery } from 'gatsby'

beforeEach(() => {
	// @ts-ignore
	StaticQuery.mockImplementationOnce((mock: any) =>
		mock.render({
			allWcProduct: {
				edges: [
					{
						node: {
							id: '367dd45d-e237-591c-9da7-c29d657da816',
							name: 'Honeymoon',
							slug: 'honeymoon',
							date_created_gmt: '2018-10-10T14:47:44',
							images: []
						}
					},
					{
						node: {
							id: '202eca74-fc90-56e7-8269-b59f18a19194',
							name: 'Watercolor texture kit Vol. 1',
							slug: 'watercolor-texture-kit-vol-1',
							date_created_gmt: '2018-09-25T20:35:51',
							images: [
								{
									id: 13,
									fullSize: {
										url: 'http://shopeverytuesday.local/wp-content/uploads/2018/06/seamless-watercolor-patterns-preview.jpg'
									},
									localFile: {
										size: 88519,
										name: 'seamless-watercolor-patterns-preview',
										id: 'c8aeed2f-ab4f-5478-adad-c12d6e4e57f3',
										childImageSharp: {
											fluid: {
												sizes: '(max-width: 1000px) 100vw, 1000px',
												src: '/static/a5ca9e682952d79588dee87feeec37a5/1a92f/seamless-watercolor-patterns-preview.jpg',
												presentationHeight: 509,
												presentationWidth: 1000
											},
											fixed: {
												width: 300,
												height: 153,
												src: '/static/a5ca9e682952d79588dee87feeec37a5/ec435/seamless-watercolor-patterns-preview.jpg'
											}
										}
									}
								}
							]
						}
					}
				]
			}
		})
	)
})
afterEach(cleanup)

describe('Product List Layout', () => {

	it('renders correctly', () => {
		const tree = renderer
			.create(
				<ProductsListLayout/>
			)
			.toJSON()
		expect(tree).toMatchSnapshot()
	})

	it('Should render correct items', () => {
		const modalRender = render(<ProductsListLayout/>)
		const list = modalRender.getByTestId('productList')
		expect(list.children.length).toEqual(2)
	})

})