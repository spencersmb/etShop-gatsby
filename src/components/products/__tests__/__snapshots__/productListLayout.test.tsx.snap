// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product List Layout renders correctly 1`] = `
<div
  className="sc-htoDjs dUAVdz"
  data-testid="productList"
>
  <div
    className="sc-bdVaJa emmDvk"
  >
    <a
      href="/products/watercolor-texture-kit-vol-1"
    >
      <div
        className="sc-bwzfXH djTlgz"
      >
        <div
          className="ListItemTop sc-bxivhb cXSKfV"
        >
          <div
            className=" gatsby-image-wrapper"
            style={
              Object {
                "overflow": "hidden",
                "position": "relative",
              }
            }
          >
            <div
              style={
                Object {
                  "paddingBottom": "0.08103727714748785%",
                  "width": "100%",
                }
              }
            />
            <img
              alt=""
              src="base64"
              style={
                Object {
                  "height": "100%",
                  "left": 0,
                  "objectFit": "cover",
                  "objectPosition": "center",
                  "opacity": 1,
                  "position": "absolute",
                  "top": 0,
                  "transitionDelay": "0.5s",
                  "width": "100%",
                }
              }
            />
            <picture>
              <img
                alt=""
                onLoad={[Function]}
                sizes="(max-width: 1000px) 100vw, 1000px"
                src="/static/a5ca9e682952d79588dee87feeec37a5/1a92f/seamless-watercolor-patterns-preview.jpg"
                srcSet="srcSet"
                style={
                  Object {
                    "height": "100%",
                    "left": 0,
                    "objectFit": "cover",
                    "objectPosition": "center",
                    "opacity": 0,
                    "position": "absolute",
                    "top": 0,
                    "transition": "opacity 0.5s",
                    "width": "100%",
                  }
                }
              />
            </picture>
            <noscript
              dangerouslySetInnerHTML={
                Object {
                  "__html": "<picture><img sizes=\\"(max-width: 1000px) 100vw, 1000px\\" srcset=\\"srcSet\\" src=\\"/static/a5ca9e682952d79588dee87feeec37a5/1a92f/seamless-watercolor-patterns-preview.jpg\\" alt=\\"\\" style=\\"position:absolute;top:0;left:0;opacity:1;width:100%;height:100%;object-fit:cover;object-position:center\\"/></picture>",
                }
              }
            />
          </div>
          <div
            className="sc-htpNat dVHrgR"
          >
            <h2>
              Watercolor texture kit Vol. 1
            </h2>
          </div>
        </div>
        <div
          className="ListItemFooter sc-ifAKCX gYZfIE"
        >
          <span />
          <p>
            $
          </p>
          <test-file-stub />
        </div>
      </div>
    </a>
  </div>
  <div
    className="sc-bdVaJa emmDvk"
  >
    <a
      href="/products/honeymoon"
    >
      <div
        className="sc-bwzfXH djTlgz"
      >
        <div
          className="ListItemTop sc-bxivhb cXSKfV"
        >
          <div
            className="sc-htpNat dVHrgR"
          >
            <h2>
              Honeymoon
            </h2>
          </div>
        </div>
        <div
          className="ListItemFooter sc-ifAKCX gYZfIE"
        >
          <span />
          <p>
            $
          </p>
          <test-file-stub />
        </div>
      </div>
    </a>
  </div>
</div>
`;
